File: ImperatorToCK3Tests/ImperatorWorldTests/Genes/AccessoryGeneTests.cpp
#include "gtest/gtest.h"
#include "../ImperatorToCK3/Source/Imperator/Genes/AccessoryGene.h"
#include <sstream>


TEST(ImperatorWorld_AccessoryGeneTests, indexCanBeSet)
{
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "\tindex=69";
	input << "}";

	const Imperator::AccessoryGene theGene(input);

	ASSERT_EQ(69, theGene.getIndex());
}

TEST(ImperatorWorld_AccessoryGeneTests, indexDefaultsTo0)
{
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const Imperator::AccessoryGene theGene(input);

	ASSERT_EQ(0, theGene.getIndex());
}

TEST(ImperatorWorld_AccessoryGeneTests, geneTemplatesDefaultToEmpty)
{
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const Imperator::AccessoryGene theGene(input);

	ASSERT_TRUE(theGene.getGeneTemplates().empty());
}

TEST(ImperatorWorld_AccessoryGeneTests, accessoryGeneIsProperlyLoaded)
{
	std::stringstream input;
	input << " = {\n";
	input << "	index = 95\n";
	input << "	inheritable = no\n";
	input << "	nerdy_hairstyles = {\n";
	input << "			index = 0\n";
	input << "			male = {\n";
	input << "				6 = male_hair_roman_5\n";
	input << "				1 = empty\n";
	input << "			}\n";
	input << "			female = {\n";
	input << "				1 = female_hair_roman_1\n";
	input << "				1 = female_hair_roman_5\n";
	input << "			}\n";
	input << "			boy = male\n";
	input << "			girl = {\n";
	input << "				1 = female_hair_roman_1\n";
	input << "				1 = female_hair_roman_5\n";
	input << "			}\n";
	input << "	}\n";
	input << "	punk_hairstyles = {\n";
	input << "		index = 1\n";
	input << "		male = {\n";
	input << "				6 = male_hair_roman_1\n";
	input << "				1 = empty\n";
	input << "		}\n";
	input << "		female = {\n";
	input << "				1 = female_hair_roman_1\n";
	input << "				1 = female_hair_roman_2\n";
	input << "		}\n";
	input << "		girl = female\n";
	input << "	} \n";
	input << "}\n";

	const Imperator::AccessoryGene gene(input);

	ASSERT_EQ(95, gene.getIndex());
	ASSERT_FALSE(gene.isInheritable());
	ASSERT_EQ(2, gene.getGeneTemplates().size());
	ASSERT_EQ(1, gene.getGeneTemplates().find("punk_hairstyles")->second.getIndex());
	ASSERT_EQ(4, gene.getGeneTemplates().find("nerdy_hairstyles")->second.getAgeSexWeightBlocs().size());
	ASSERT_EQ(3, gene.getGeneTemplates().find("punk_hairstyles")->second.getAgeSexWeightBlocs().size());
}

File: ImperatorToCK3.UnitTests/Imperator/Genes/AccessoryGeneTests.cs
using commonItems;
using ImperatorToCK3.Imperator.Genes;
using Xunit;

namespace ImperatorToCK3.UnitTests.Imperator.Genes {
	public class AccessoryGeneTests {

		[Fact]
		public void indexCanBeSet() {
			var reader = new BufferedReader(
				"=\n" +
				"{\n" +
				"\tindex=69" +
				"}"
			);
			var gene = new AccessoryGene(reader);

			Assert.Equal((uint)69, gene.Index);
		}

		[Fact]
		public void indexDefaultsTo0() {
			var reader = new BufferedReader("={}");
			var gene = new AccessoryGene(reader);

			Assert.Equal((uint)0, gene.Index);
		}

		[Fact]
		public void geneTemplatesDefaultToEmpty() {
			var reader = new BufferedReader("={}");
			var gene = new AccessoryGene(reader);

			Assert.Empty(gene.GeneTemplates);
		}

		[Fact]
		public void accessoryGeneIsProperlyLoaded() {
			var reader = new BufferedReader(
				" = {\n" +
				"	index = 95\n" +
				"	inheritable = no\n" +
				"	nerdy_hairstyles = {\n" +
				"			index = 0\n" +
				"			male = {\n" +
				"				6 = male_hair_roman_5\n" +
				"				1 = empty\n" +
				"			}\n" +
				"			female = {\n" +
				"				1 = female_hair_roman_1\n" +
				"				1 = female_hair_roman_5\n" +
				"			}\n" +
				"			boy = male\n" +
				"			girl = {\n" +
				"				1 = female_hair_roman_1\n" +
				"				1 = female_hair_roman_5\n" +
				"			}\n" +
				"	}\n" +
				"	punk_hairstyles = {\n" +
				"		index = 1\n" +
				"		male = {\n" +
				"				6 = male_hair_roman_1\n" +
				"				1 = empty\n" +
				"		}\n" +
				"		female = {\n" +
				"				1 = female_hair_roman_1\n" +
				"				1 = female_hair_roman_2\n" +
				"		}\n" +
				"		girl = female\n" +
				"	} \n" +
				"}\n"
			);
			var gene = new AccessoryGene(reader);

			Assert.Equal((uint)95, gene.Index);
			Assert.False(gene.Inheritable);
			Assert.Equal(2, gene.GeneTemplates.Count);
			Assert.Equal((uint)1, gene.GeneTemplates["punk_hairstyles"].Index);
			Assert.Equal(4, gene.GeneTemplates["nerdy_hairstyles"].AgeSexWeightBlocks.Count);
			Assert.Equal(3, gene.GeneTemplates["punk_hairstyles"].AgeSexWeightBlocks.Count);
		}
	}
}
